Blog: How we do trunk-based development (and why you should too) (#7570)
PostHog/posthog.com: contents/product-engineers/trunk-based-development.md

---
title: How we do trunk-based development (and why you should too)
date: 2024-01-19
author: ["ian-vanagas"]
rootpage: /blog
featuredImage: ../images/blog/posthog-engineering-blog.png
featuredImageType: full
tags:
 - Product engineers
 - Engineering
 - Feature management
---

What do trees, elephants, and successful software all have in common? They all have strong trunks helping them grow big and strong. In the case of software, the trunk is the main codebase engineers make changes to and that deploys to users. 

There are many strategies for branching off and making changes to this trunk. The one we use at PostHog is trunk-based development. It helps us reduce work-in-progress, ship faster, keep momentum, and ultimately, build a more successful product. 

> This post covers what trunk-based development is, its benefits, why we do it at PostHog, and how you can implement it for yourself.

Reviewer: Couple of thoughts:

- Could we make this intro more opinionated? Maybe talk about some of the benefits to us – talk up our shipping speed etc.
- Add more context about PostHog. Who are we? Why should the reader listen to us?

Additional... I feel like the reader hear is an engineer / engineering leader who want to enact change. How can we make the intro + headline speak to them?

## Why and how PostHog does trunk-based development

PostHog does trunk-based development. The big reason why is to **minimize work-in-progress**. This is a [core philosophy of our development process](/handbook/engineering/development-process#sizing-tasks-and-reducing-wip) at PostHog and relates to all three of the benefits of the last section.

Implementing this philosophy (and trunk-based development) means:

- Optimizing pull requests for quality of implementation *and* the speed of merge.

- Keeping changes doable in one day, including code review and QA.

- If not, split them into smaller changes until they fit into one day.

- Responding to colleagues' review requests early in your day to help unblock them.

- Pushing code out early. PRs are better than issues, even if they are half-done. This enables more concrete feedback on the approach.

- Merging whenever. Our testing, reviewing, and build processes should be good enough that we are confident doing this.

As proof of this, our median pull request age when merged is **just under 8 hours**. On an average day, we merge around 20 pull requests in our main repo.

> All of this helps us get the benefits trunk-based development provides. It might seem like a lot. Pushing code early and merging whenever may seem especially scary, but we have two tools that help us:

Reviewer: Could you take a second stab at these two paras? I find them difficult to parse atm.

### 1. Test-driven development

> PostHog expects engineers to write tests for their code as they are working on it and before they merge it. They run these locally then add them to our test suite that runs on every PR.

Reviewer: ```suggestion
We expect engineers to write tests for their code as they are working on it and before they merge it. They run these locally then add them to our test suite that runs on every PR.
```

![Checks](../images/blog/trunk-based-development/checks.png)

You can look at [a pull request on our repo](https://github.com/PostHog/posthog/pulls) to see the suite of end-to-end, frontend, linting, Storybook, code quality, visual regression, and migration tests before merging. This enables us to be confident in the code we are merging.

> ### 2. Feature flags
> 
> [Feature flags](/feature-flags) enable us to include code in the "trunk" but not release it to users. This [decouples deployment from release](/product-engineers/decouple-deployment-from-release). 
> 
> Once deployed, we can then test in production, dogfood, beta test, get feedback, and progressively roll out the feature. We currently have [67 feature flags](https://github.com/PostHog/posthog/blob/249d82043bcd70bb3e95a0553551246998d0d160/frontend/src/lib/constants.tsx#L125C1-L198C3) in various stages of release in our main PostHog repo.

Reviewer: ```suggestion
### 2. Feature flags

[Feature flags](/feature-flags) enable us to include code in the "trunk" without immediately releasing it to users. This [decouples deployment from release](/product-engineers/decouple-deployment-from-release). 

Once deployed, we can then test in production, dogfood, beta test, get feedback, and progressively roll out the feature. 

We currently have [67 feature flags](https://github.com/PostHog/posthog/blob/249d82043bcd70bb3e95a0553551246998d0d160/frontend/src/lib/constants.tsx#L125C1-L198C3) in various stages of release in our main PostHog repo.
```

## How you can implement trunk-based development

> Implementing trunk-based development starts with a philosophy shift. Teams must encourage (or force) developers towards short-lived branches rather than long-lived ones. This means breaking up PRs into smaller changes, prioritizing reviews, and pushing code out earlier. 

Reviewer: ```suggestion
Implementing trunk-based development starts with a philosophy shift. Teams must encourage (or force) developers towards short-lived branches, rather than long-lived ones. This means breaking up PRs into smaller changes, prioritizing reviews, and pushing code out earlier. 
```

> ---
> title: How we do trunk-based development (and why you should too)
> date: 2024-01-30
> author: ["ian-vanagas"]
> rootpage: /blog
> featuredImage: ../images/blog/posthog-engineering-blog.png
> featuredImageType: full
> tags:
>  - Product engineers
>  - Engineering
>  - Feature management
> ---

Reviewer: ```suggestion
---
title: How we do trunk-based development (and why you should too)
date: 2024-02-05
author: ["ian-vanagas"]
rootpage: /blog
featuredImage: ../images/blog/posthog-engineering-blog.png
featuredImageType: full
tags:
 - Engineering
 - Feature management
---
```
